File: 3312-sorted-gcd-pair-queries/3312-sorted-gcd-pair-queries.java
class Solution {
    public int[] gcdValues(int[] nums, long[] queries) {
        int maxVal = 0;
        for (int num : nums) {
            maxVal = Math.max(maxVal, num);
        }
        int[] count = new int[maxVal + 1];
        for (int num : nums) {
            count[num]++;
        }
        long[] exactCount = new long[maxVal + 1];
        for (int i = maxVal; i >= 1; i--) {
            long multiplesCount = 0;
            for (int j = i; j <= maxVal; j += i) {
                multiplesCount += count[j];
            }
            long pairs = (multiplesCount * (multiplesCount - 1)) / 2;
            for (int j = 2 * i; j <= maxVal; j += i) {
                pairs -= exactCount[j];
            }
            exactCount[i] = pairs;
        }
        long[] prefixSum = new long[maxVal + 1];
        for (int i = 1; i <= maxVal; i++) {
            prefixSum[i] = prefixSum[i - 1] + exactCount[i];
        }
        int[] result = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            long target = queries[i];
            int left = 1, right = maxVal;
            int ans = maxVal;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (prefixSum[mid] > target) {
                    ans = mid;
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            result[i] = ans;
        }
        return result;
    }
}

File: 3312-sorted-gcd-pair-queries/3312-sorted-gcd-pair-queries.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> gcdValues(vector<int>& nums, vector<long long>& queries) {
        int maxVal = 0;
        for (int num : nums) {
            maxVal = max(maxVal, num);
        }
        vector<int> count(maxVal + 1, 0);
        for (int num : nums) {
            count[num]++;
        }
        vector<long long> exactCount(maxVal + 1, 0);
        for (int i = maxVal; i >= 1; i--) {
            long long multiplesCount = 0;
            for (int j = i; j <= maxVal; j += i) {
                multiplesCount += count[j];
            }
            long long pairs = (multiplesCount * (multiplesCount - 1)) / 2;
            for (int j = 2 * i; j <= maxVal; j += i) {
                pairs -= exactCount[j];
            }
            
            exactCount[i] = pairs;
        }
        vector<long long> prefixSum(maxVal + 1, 0);
        for (int i = 1; i <= maxVal; i++) {
            prefixSum[i] = prefixSum[i - 1] + exactCount[i];
        }
        vector<int> result(queries.size());
        for (int i = 0; i < queries.size(); i++) {
            auto it = upper_bound(prefixSum.begin() + 1, prefixSum.end(), queries[i]);
            result[i] = distance(prefixSum.begin(), it);
        }
        return result;
    }
};